### ISSUE TITLE
[core] Tree traversal revision

### ISSUE THREAD

**Issue by jsotuyod** [pmd/pmd#904]
I was recently digging into #370, which is false reporting on lambdas when finding code such as:

```java
LOGGER.info(() -> "Bla " + " bla");
```

 when I realized the problem laid on this line

https://github.com/pmd/pmd/blob/cd9f4f2248f4358bfc6e1acf979102480ec816dd/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/GuardLogStatementRule.java#L97

`hasDescendantOfType` completely ignores find boundaries. `getFirstDescendantOfType` also completely ignores find boundaries. Moreover, **all** tree transversal operations ignore find boundaries, except for `findDescendantsOfType` which doesn't (but has a convoluted overload which allows to).

So, `node.hasDescendantOfType(XXX.class)` may return true, `node.getFirstDescendantOfType(XXX.class)` won't return null, but `node.findDescendantsOfType(XXX.class)` may return empty. Rejoyce!

Moreover, most often than not, we don't really want to transverse along find boundaries. Actually, pretty much all current rules are broken on this respect. Some brief examples I was able to find in under 15'

Apparently when finding

```
public void foo() {
  class Bar {
    private Object o = new Object();
  }
}
```

The whole `class Bar { … }` block is a single `ASTBlockStatement` which is convinced this is an allocation statement.

https://github.com/pmd/pmd/blob/cd9f4f2248f4358bfc6e1acf979102480ec816dd/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTBlockStatement.java#L30-L32

But lambdas in particular are easier to find anywhere and completely break logic:

The complexity of the statement `return () -> { someValue && doStuff() };` is 1, even though we are simply creating a lambda and in doing so performing exactly 0 boolean operations.

https://github.com/pmd/pmd/blob/cd9f4f2248f4358bfc6e1acf979102480ec816dd/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/CycloMetric.java#L62-L68

Returning a lambda can also wreck havoc on sun secure rules

https://github.com/pmd/pmd/blob/cd9f4f2248f4358bfc6e1acf979102480ec816dd/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/AbstractSunSecureRule.java#L82-L93

I have so far been unable to find a single case where:
 - we don't intend to look **ONLY within** find boundaries
 - it's irrelevant due to the AST format (ie: looking for imports, which can only be direct children of the ACU)

On the first scenario, the current transversing is broken and may produce FPs / FNs. We may actually be reporting duplicate violations as lambdas / nested local classes statements may be analyzed more than once.
On the second scenario, we are simply inefficient by digging too deep

I did a quick test, and changing `getFirstDescendantOfType` (and therefore `hasDescendantOfType`) to respect find boundaries resulted in 0 errors, except for for this one statement where on a test we are being lazy:

https://github.com/pmd/pmd/blob/cd9f4f2248f4358bfc6e1acf979102480ec816dd/pmd-java/src/test/java/net/sourceforge/pmd/typeresolution/ClassTypeResolverTest.java#L237-L239

On the other hand, I'm sure I can come up with tests for all these rules where lambdas an local classes produce FPs / FNs.

@adangel @oowekyala what do you think? Shall we change the default behavior on PMD 6.1.0? Shall we deprecate those methods and add new ones allowing to cross boundaries on demand? Is there really a case for crossing boundaries?

**Comment by oowekyala** [comment]
Awesome research! I can't either find a use case for crossing find boundaries, and your examples clearly show there's a problem. I can see a possible confusion arising though, since AST visitors do cross boundaries by default. This will need to be clearly documented. 

For instance, some metrics search too deep (Cyclo, see #837, but I think nearly all others do the same, eg NPath), because they use a visitor. A general [fix](https://github.com/pmd/pmd/pull/895/commits/b30d64dd4f17510b3904cfd040ba62940b871f9f#diff-7a333e2e9d3f6bbe23270df2521e9d33R44) is difficult to share without defining a new visitor adapter class. But this problem is mostly confined to metrics, not to rules.

To come back on those AbstractNode methods, I think making them default to not crossing boundaries would be fine, and we could provide overloads to cover all possibilities, just in case. Uniformising their names and interface would also be nice, since some start with "find", others with "get". The weird overload where you pass the result list as parameter should definitely be deprecated, or at least its public use. The changeset should be fine for 6.1.0

**Comment by oowekyala** [comment]
I just ran into another bug caused by this. In the following:
```java
box.add(new JButton(new AbstractAction("expand  ") {
            @Override public void actionPerformed(ActionEvent e) {
                TreeNode root = (TreeNode) tree.getModel().getRoot();
                visitAll(tree, new TreePath(root), true); // line 14
            }
}));
```
there are two ASTAllocationExpression, and only the inner one is an anonymous class. However, the ClassOrInterfaceType of the outer one returns true for `isAnonymousClass`, because that method is defined by
```java
 public boolean isAnonymousClass() {
     return jjtGetParent().hasDescendantOfType(ASTClassOrInterfaceBody.class);
}
```
which goes too deep...

**Comment by jsotuyod** [comment]
@adangel I'd love to hear your input on this. I'm too inclined to get this done, but want to make sure you don't see any plausible issues with this behavior change.

**Comment by adangel** [comment]
@jsotuyod 

I'm definitely in favor for fixing this:

> So, node.hasDescendantOfType(XXX.class) may return true,
> node.getFirstDescendantOfType(XXX.class) won't return null, but
> node.findDescendantsOfType(XXX.class) may return empty. Rejoyce!

This clearly violations the principle of least surprise....


Unfortunately, `Node.isFindBoundary()` has no documentation and we seem to have no documentation about this concept at all (at least, a search for "boundary" in docs yields empty).
It's actually quite an important concept for the AST traversal of a language - and it's language dependent (not needed for all languages), so the default implementation in AbstractNode can stay as-is and simply return `false`.

I think, yes, we should change the default behavior to always respect the find boundary.
I do not think, we need to deprecate these methods, we would just change the behavior. Reading your research, I'm convinced, that this new behavior (respecting find boundaries) is more correct.
We should not add overloaded methods, which can control the find boundary, just now - only as soon as we need them for a case, where we need to cross the boundaries.

As for the method names, I don't think, they are too bad: `find*` returns a List, `getFirst*` returns a single result and `has*` return a boolean.
There is one method however: `getParentsOfType(...)` which we should probably deprecate and change to `findParentsOfType(...)`.

There is one more thing to consider: XPath... As far as I see, both Jaxen and Saxon child axis iterators don't check for the boundary, but we could change this in a second step.
